pkg/process/process.go

```diff
@@ -0,0 +1,234 @@
+// TODO(kakkoyun): This is not unique enough. We will need to add more fields.
+// For example, we can use the process start time to make sure we are not.
+// However, reading the stats file is expensive. For now, we will rely on the PID.
+// I will keep the struct to make it easier to add more fields in the future.
+type key struct {
+	pid int
+}
+
+type Tree struct {
+	tree        map[key]*process
+	mtx         *sync.RWMutex
+	disappeared map[key]struct{}
+	reused      map[key]*process
+}
+
+type process struct {
+	procfs.Proc
+
+	tree *Tree
+
+	parent int
+	// name of the field is the same as the one in kernel struct.
+	starttime uint64
+}
+
+func (p *process) ancestors() []*procfs.Proc {
+	p.tree.mtx.RLock()
+	defer p.tree.mtx.RUnlock()
+
+	ancestors := []*procfs.Proc{}
+	proc := p
+	for {
+		if proc.parent == 0 {
+			break
+		}
+		k := key{pid: proc.parent}
+		parent, ok := p.tree.tree[k]
+		if !ok {
+			break
+		}
+		ancestors = append(ancestors, &parent.Proc)
+		proc = parent
+	}
+	return ancestors
+}
+
+// NewTree returns a new process tree with current state of all the processes on the system.
+func NewTree(ctx context.Context, resetDuration time.Duration) *Tree {
+	t := &Tree{
+		tree:        make(map[key]*process),
+		disappeared: make(map[key]struct{}),
+		mtx:         &sync.RWMutex{},
+	}
+	go func() {
+		// This is a very naive pruning implementation. It will be improved in the future.
+		ticker := time.NewTicker(resetDuration)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				// Prune disappeared processes.
+				t.mtx.Lock()
+				for k := range t.disappeared {
+					delete(t.tree, k)
+				}
+				t.mtx.Unlock()
+				t.disappeared = make(map[key]struct{})
+
+				t.mtx.RLock()
+				for k, p := range t.tree {
+					proc, err := procfs.NewProc(p.PID)
+					if err != nil {
+						// Will be pruned in the next iteration.
+						t.disappeared[k] = struct{}{}
+						continue
+					}
+					stat, err := proc.Stat()
+					if err != nil {
+						// Will be pruned in the next iteration.
+						t.disappeared[k] = struct{}{}
+						continue
+					}
+					// It is possible that the process has been reused. Update the process.
+					if p.starttime != stat.Starttime {
+						t.reused[k] = &process{
+							Proc:      proc,
+							tree:      t,
+							parent:    stat.PPID,
+							starttime: stat.Starttime,
+						}
+						continue
+					}
+				}
+				t.mtx.RUnlock()
+
+				// Update the process tree.
+				t.mtx.Lock()
+				for k, p := range t.reused {
+					t.tree[k] = p
+				}
+				t.mtx.Unlock()
+				t.reused = make(map[key]*process)
+			}
+		}
+	}()
+	return t
+}
+
+// FindAllAncestorProcessIDsInSameCgroup returns all ancestor process IDs for a given PID in the same cgroup.
+func (t *Tree) FindAllAncestorProcessIDsInSameCgroup(pid int) ([]int, error) {
+	// Fast path. Find the process if it exists in the process tree.
+	k := key{pid: pid}
+	t.mtx.RLock()
```

Comment:
Can we move the part from here to the read-unlock to a separate function, that returns `[]int, needsRefresh`? We also only need to lock around the `t.tree[k]` access, the procfs accesses can theoretically block arbitrarily long but it actually doesn't matter for the read lock.

Reply:
Actually this could use the same `Get` function suggested elsewhere.